### PR TITLE
feat: find dialog 1760

### PR DESCRIPTION
**Pull request by MayurSMahajan**
This PR aims at creating a FInd Dialog, that is able to find string patterns in editor document. 

This PR introduces three new files for implementing this feature: 
1. `find_menu_service.dart`
2. `find_replace_widget.dart`
3. `search_service.dart`

****Find Menu Service****
This service is actually implemented by FindReplaceMenu, which is responsible for popping a Find Replace Menu onto the editor. 
It is able to do so by adding a new `OverlayEntry` to the editor.

****Find Replace Widget****
It is simply a widget component strictly used for presentation. It separates the FindMenuService from the SearchService. It describes the UI of the Find Replace Menu. 

****Search Service****
The Heavy Logic Lifting is done by a class called SearchService. This class contains public methods like: 
`findAndHighlight`: finds the pattern and highlights it.
`navigateToMatch`: navigates between matched patterns in the vertical direction.
`unHighlight`: unhighlights the matches when a new pattern is searched or the find dialog is closed.

It also contains a bunch of private utility methods, one particularly interesting method is the `_boyerMooreSearch` method, which implements the world-renowned Boyer Moore Search algorithm for searching the pattern in the editor.

### Limitations

- [x] Scrolling to the next match using arrow keys, does not behave as expected.
- [x] Behaviour on already highlighted patterns.

### Features Yet to be implemented

- [x] Replace handler
- [x] Tests

### CODE REVIEW

File: lib/src/render/find_replace_menu/search_service.dart [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+      offset: start.offset + queriedPattern.length,
+    );
+
+    editorState.updateCursorSelection(Selection(start: start, end: end));
+  }
+
+  //This is a standard algorithm used for searching patterns in long text samples
+  //It is more efficient than brute force searching because it is able to skip
+  //characters that will never possibly match with required pattern.
+  List<int> _boyerMooreSearch(String pattern, String text) {
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/search_service.dart:177]: If we need to maintain an algorithm like this, then it should probably be managed a bit more clearly. Like having its own class.

At least I would prefer that personally.

File: test/render/find_replace_menu/find_replace_menu_find_test.dart [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+      const pattern = 'Welcome';
+      const closeBtnKey = Key('closeButton');
+
+      final editor = await _prepareWithTextInputForFind(
+        tester,
+        lines: 3,
+        pattern: pattern,
+      );
+
+      var selection =
```
Comment (Xazin) [inline test/render/find_replace_menu/find_replace_menu_find_test.dart:222]: Final

File: lib/src/render/find_replace_menu/search_service.dart (hunk 1 above)
Comment (MayurSMahajan) [reply lib/src/render/find_replace_menu/search_service.dart:177]: I also considered putting the algorithm in a separate class, but then decided against it. Putting it in a separate class makes more sense though.

File: lib/src/render/find_replace_menu/search_service.dart [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        }
+        //finally we will highlight all the mathces.
+        _highlightMatches(n.path, matches, pattern.length);
+      }
+    }
+
+    selectedIndex = matchedPositions.length - 1;
+  }
+
+  void unHighlight(String pattern) {
```
[the remark this answers is not in the document]
Comment (MayurSMahajan) [reply lib/src/render/find_replace_menu/search_service.dart:66]: If we remove this method, then some of my tests are failing, also keeping this method improves readability, the reader knows when we are searching and highlighting and when we are unhighlighting. In the future, we may want to add more logic for unhighlighting depending on how we change the highlighting feature now. I would say we keep the method. But if you still insist, then we can remove it, your call.

[the remark this answers is not in the document]
Comment (MayurSMahajan) [reply lib/src/render/find_replace_menu/search_service.dart]: Okay I will remove it.

**Comment by MayurSMahajan** [comment]
@Xazin I wanted to ask you this: Currently I am using the default highlight color which comes from the editor styles to highlight found matches, should I use a different color instead?

**Review by a-wallen — COMMENTED** [review]
Very nice, I'm stoked for this. Could you add a handful of unit tests for the search algorithm?

**Comment by MayurSMahajan** [comment]
Sure Alex!

**Comment by Xazin** [comment]
> @Xazin I wanted to ask you this: Currently I am using the default highlight color which comes from the editor styles to highlight found matches, should I use a different color instead?

Yes, let's expose 2 colors for the SearchStyle, one for highlighting results, and one for highlighting the current result (for traversal)

File: lib/src/history/undo_manager.dart [hunk 4]
```diff
@@ -149,4 +149,12 @@ class UndoManager {
       ),
     );
   }
+
+  void forgetRecentUndo() {
+    Log.editor.debug('forgetRecentUndo');
+    if (state == null) {
+      return;
+    }
+    undoStack.pop();
```
Comment (Xazin) [inline lib/src/history/undo_manager.dart:158]:
```suggestion
    if (state != null) {
      undoStack.pop();
    }
```

File: lib/src/render/find_replace_menu/find_menu_service.dart [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  final bool replaceFlag;
+  final double topOffset = 52;
+  final double rightOffset = 40;
+
+  OverlayEntry? _findReplaceMenuEntry;
+  bool _selectionUpdateByInner = false;
+
+  @override
+  void dismiss() {
+    if (_findReplaceMenuEntry != null) {
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_menu_service.dart:29]: No need to go through all of this logic if for some reason `dismiss()` is called without there being a current entry.
```suggestion
    if (_findReplaceMenuEntry == null) {
      return;
    }

    if (_findReplaceMenuEntry != null) {
```

File: lib/src/render/find_replace_menu/find_replace_widget.dart [hunk 6]
```diff
@@ -0,0 +1,148 @@
+import 'package:appflowy_editor/appflowy_editor.dart';
+import 'package:appflowy_editor/src/render/find_replace_menu/search_service.dart';
+import 'package:flutter/material.dart';
+
+class FindMenuWidget extends StatefulWidget {
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_replace_widget.dart:5]: I believe the FindMenuWidget to be too big and chunky, I've left some comments on how to address this.

File: lib/src/render/find_replace_menu/find_replace_widget.dart [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+              }),
+              icon: replaceFlag
+                  ? const Icon(Icons.expand_less)
+                  : const Icon(Icons.expand_more),
+            ),
+            Padding(
+              padding: const EdgeInsets.all(6.0),
+              child: SizedBox(
+                width: 200,
+                height: 50,
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_replace_widget.dart:55]:
```suggestion
                height: 30,
```

File: lib/src/render/find_replace_menu/find_replace_widget.dart [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+                child: TextField(
+                  key: const Key('findTextField'),
+                  autofocus: true,
+                  controller: findController,
+                  onSubmitted: (_) => _searchPattern(),
+                  decoration: const InputDecoration(
+                    border: OutlineInputBorder(),
+                    hintText: 'Enter text to search',
+                  ),
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_replace_widget.dart:64]:
```suggestion
                  decoration: const InputDecoration(
                    contentPadding: EdgeInsets.symmetric(horizontal: 8),
                    border: OutlineInputBorder(),
                    hintText: 'Enter text to search',
                  ),
```

File: lib/src/render/find_replace_menu/find_replace_widget.dart [hunk 9]
```diff
@@ -0,0 +1,148 @@
+import 'package:appflowy_editor/appflowy_editor.dart';
+import 'package:appflowy_editor/src/render/find_replace_menu/search_service.dart';
+import 'package:flutter/material.dart';
+
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_replace_widget.dart:4]:
```suggestion

const double _iconSize = 20;

```

File: lib/src/render/find_replace_menu/find_replace_widget.dart [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+                  onSubmitted: (_) => _searchPattern(),
+                  decoration: const InputDecoration(
+                    border: OutlineInputBorder(),
+                    hintText: 'Enter text to search',
+                  ),
+                ),
+              ),
+            ),
+            IconButton(
+              key: const Key('previousMatchButton'),
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_replace_widget.dart:69]:
```suggestion
              key: const Key('previousMatchButton'),
              iconSize: _iconSize,
```

File: lib/src/render/find_replace_menu/find_replace_widget.dart [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+              onPressed: () => searchService.navigateToMatch(moveUp: true),
+              icon: const Icon(Icons.arrow_upward),
+              tooltip: 'Previous Match',
+            ),
+            IconButton(
+              key: const Key('nextMatchButton'),
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_replace_widget.dart:75]:
```suggestion
              key: const Key('nextMatchButton'),
              iconSize: _iconSize,
```

File: lib/src/render/find_replace_menu/find_replace_widget.dart [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+              onPressed: () => searchService.navigateToMatch(),
+              icon: const Icon(Icons.arrow_downward),
+              tooltip: 'Next Match',
+            ),
+            IconButton(
+              key: const Key('closeButton'),
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_replace_widget.dart:81]:
```suggestion
              key: const Key('closeButton'),
              iconSize: _iconSize,
```

File: lib/src/render/find_replace_menu/find_replace_widget.dart [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+          ],
+        ),
+        replaceFlag
+            ? Row(
+                children: [
+                  Padding(
+                    padding: const EdgeInsets.all(6.0),
+                    child: SizedBox(
+                      width: 200,
+                      height: 50,
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_replace_widget.dart:99]:
```suggestion
                      height: 30,
```

File: lib/src/render/find_replace_menu/find_replace_widget.dart [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+          children: [
+            IconButton(
+              onPressed: () => setState(() {
+                replaceFlag = !replaceFlag;
+              }),
+              icon: replaceFlag
+                  ? const Icon(Icons.expand_less)
+                  : const Icon(Icons.expand_more),
+            ),
+            Padding(
```
Comment (Xazin) [inline lib/src/render/find_replace_menu/find_replace_widget.dart:51]: Remove Padding widget

**Review by Xazin — COMMENTED** [review]
There also seems to be a bug, where if I press `CTRL + F` multiple times, then multiple of the find/replace menu opens up on top of each other.

**Comment by MayurSMahajan** [comment]
@Xazin, the core feature is complete with this set of commits, however, there are two issues to address:
1. I am using a constant color for highlighting, we want to be able to use a customizable color instead, currently, @LucasXu0 and I are discussing the same.
2. When multiple matches are found, the scrolling speed while navigating between these found matches is very slow, this is something I would you to take a look at, if possible.

**Comment by Xazin** [comment]
> @Xazin, the core feature is complete with this set of commits, however, there are two issues to address:
> 
> 1. I am using a constant color for highlighting, we want to be able to use a customizable color instead, currently, @LucasXu0 and I are discussing the same.
> 
> 2. When multiple matches are found, the scrolling speed while navigating between these found matches is very slow, this is something I would you to take a look at, if possible.

Hey Mayur, thanks for the update.

If possible leave these two issues to me, I will take a look on Monday and give you an update. 👌

**Comment by MayurSMahajan** [comment]
> > @Xazin, the core feature is complete with this set of commits, however, there are two issues to address:
> > 
> > 1. I am using a constant color for highlighting, we want to be able to use a customizable color instead, currently, @LucasXu0 and I are discussing the same.
> > 2. When multiple matches are found, the scrolling speed while navigating between these found matches is very slow, this is something I would you to take a look at, if possible.
> 
> Hey Mayur, thanks for the update.
> 
> If possible leave these two issues to me, I will take a look on Monday and give you an update. 👌

Sure!

File: lib/src/editor/find_replace_menu/search_service.dart [hunk 15]
```diff
@@ -104,34 +115,40 @@ class SearchService {
...
-    editorState.updateSelectionWithReason(
+    await editorState.updateSelectionWithReason(
       Selection(start: start, end: end),
       reason: isNavigating
           ? SelectionUpdateReason.searchNavigate
           : SelectionUpdateReason.searchHighlight,
     );
+
+    editorState.service.keyboardService?.disable();
+    editorState.service.selectionService.clearCursor();
```
Comment (Xazin) [inline lib/src/editor/find_replace_menu/search_service.dart:148]: I guess this was a remnant of my testing, we can remove

File: lib/src/editor/find_replace_menu/search_service.dart [hunk 16]
```diff
@@ -104,34 +115,40 @@ class SearchService {
...
   }
 
   /// This method takes in a boolean parameter moveUp, if set to true,
   /// the match located above the current selected match is newly selected.
   /// Otherwise the match below the current selected match is newly selected.
-  void navigateToMatch({bool moveUp = false}) {
+  void navigateToMatch({bool moveUp = false, bool keepFocus = false}) {
```
Comment (Xazin) [inline lib/src/editor/find_replace_menu/search_service.dart:154]: Same with the keepFocus, we can remove.

I was looking at the way the focus is lost from the search field when navigating to a result.

File: lib/src/editor/editor_component/service/shortcuts/command_shortcut_events/find_replace_command.dart [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+  // // delete the selection
+  // if (!selection.isCollapsed) {
+  //   await editorState.deleteSelection(selection);
+  // }
+
+  final afterSelection = editorState.selection;
+  if (afterSelection == null || !afterSelection.isCollapsed) {
+    assert(false, 'the selection should be collapsed');
+    return KeyEventResult.handled;
+  }
```
Comment (Xazin) [inline lib/src/editor/editor_component/service/shortcuts/command_shortcut_events/find_replace_command.dart:112]: We don't actually care about the selection do we? If not we can remove all of these checks.

And at the least remove the commented out code.

File: lib/src/editor/find_replace_menu/search_service.dart [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+    );
+
+    final selection = Selection(start: currentMatch, end: end);
+    _applySelectedHighlightColor(selection);
+
+    if (moveUp) {
+      selectedIndex =
+          selectedIndex - 1 < 0 ? matchedPositions.length - 1 : --selectedIndex;
+
+      Position match = matchedPositions[selectedIndex];
```
Comment (Xazin) [inline lib/src/editor/find_replace_menu/search_service.dart:172]:
```suggestion
      final match = matchedPositions[selectedIndex];
```

File: lib/src/editor/find_replace_menu/search_service.dart [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+      _selectWordAtPosition(match, true);
+    } else {
+      selectedIndex =
+          (selectedIndex + 1) < matchedPositions.length ? ++selectedIndex : 0;
+
+      final match = matchedPositions[selectedIndex];
+      _selectWordAtPosition(match, true);
+    }
+  }
```
Comment (Xazin) [inline lib/src/editor/find_replace_menu/search_service.dart:181]:
```suggestion
    if (moveUp) {
      selectedIndex =
          selectedIndex - 1 < 0 ? matchedPositions.length - 1 : --selectedIndex;
    } else {
      selectedIndex =
          (selectedIndex + 1) < matchedPositions.length ? ++selectedIndex : 0;
    }
    
    final match = matchedPositions[selectedIndex];
    _selectWordAtPosition(match, true);
```

File: lib/src/editor/toolbar/utils/format_color.dart [hunk 20]
```diff
@@ -1,15 +1,25 @@
...
 }
 
-void formatFontColor(EditorState editorState, String color) {
+void formatFontColor(
+  EditorState editorState,
+  Selection? selection,
+  String color,
+) {
   editorState.formatDelta(
     editorState.selection,
```
Comment (Xazin) [inline lib/src/editor/toolbar/utils/format_color.dart:21]:
```suggestion
    selection,
```

File: lib/src/editor/editor_component/service/shortcuts/command_shortcut_events/find_replace_command.dart (hunk 17 above)
Comment (MayurSMahajan) [reply lib/src/editor/editor_component/service/shortcuts/command_shortcut_events/find_replace_command.dart:112]: Yes all of this code can be removed.

**Comment by Xazin** [comment]
CC @LucasXu0

**Comment by LucasXu0** [comment]
@Xazin @MayurSMahajan Is this PR ready for merging? The CI failed.

**Comment by Xazin** [comment]
> @Xazin @MayurSMahajan Is this PR ready for merging? The CI failed.

I'm fixing it right now and will merge in 2 minutes.

**Pull request merged**